### DISCUSSION TITLE
Build device code in advance with static library

### DISCUSSION THREAD

**Discussion by Luigi-Crisci** [intel/llvm#9620] [Q&A] [answered]
Hello,
I have a cmake project with this structure:
- A static library `lib.a` with multiple kernel in different source files
- An executable `a.exe` linkinng to `lib.a`  

When targeting the NVIDIA backend, I want to set different `ptxas` options for each kernel, but when building the project the assembler is called during `a.exe` linking phase, thus the options are ignored as the executable is a single translation unit.  
Is there a way to tell DPC++ to build the device code in advance?

I was trying the same thing with OpenSYCL and it call `ptxas` while building the kernels during `lib.a` building phase, so it works smoothly.

**Comment by Luigi-Crisci** [comment]
I forgot to say that, to set the options for each kernel, I was using the cmake option `set_source_files_properties("KERNEL.cpp" [COMPILE_FLAGS|LINK_FLAGS]  "-Xcuda-ptxas OPTION")`. But it's completely ignored as the assembler is not called wen building the static library target

**Reply by AlexeySachkov** [reply, marked as answer]
Hi @Luigi-Crisci, 

Unfortunately, at the moment this functionality is not available neither for CUDA backend, nor for any other targets. The problem is in fundamental design of the toolchain: device compiler is invoked at link stage, when all device code from all translation units is visible. This was done to support `SYCL_EXTERNAL`.

I'm aware of other requests of the same functionality, i.e. a way to schedule device compiler pass _early_: either at static lib level, or even at a translation unit level. Motivation is the same as yours: have ability to specify per-translation unit backend-specific compilation flags.

AFAIK, someone is already working on that, but can't provide any timelines. Tagging @asudarsa and @mdtoguchi for awareness here.
